src/UI/ValueFactory.cs

```diff
@@ -388,5 +406,33 @@ private static bool GetNewColorSchemeValue(Design design, out object? newValue)
                 return false;
             }
         }
+
+        /// <summary>
+        /// Returns true if path is to a directory.
+        /// Thanks: https://stackoverflow.com/a/19596821/4824531
+        /// </summary>
+        /// <param name="path"></param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentNullException"></exception>
+        static bool IsPathDirectory(string path)
```

Comment:
Directory.Exists is already this, is it not? I'm not clear on what this method is for.

And has it been tested on symlinks, including on same volume, different volumes, and different systems (the latter should fail by default on windows for security reasons)?

Here are the remarks from System.IO.Directory.Exists(string):

> The path parameter is permitted to specify relative or absolute path information. Relative path information is interpreted as relative to the current working directory.
> 
> **Trailing spaces are removed from the end of the path parameter before checking whether the directory exists.**
> 
>The [Exists](https://learn.microsoft.com/en-us/dotnet/api/system.io.directory.exists?view=net-8.0) method returns false if any error occurs while trying to determine if the specified file exists. This can occur in situations that raise exceptions such as passing a file name with invalid characters or too many characters, a failing or missing disk, or if the caller does not have permission to read the file.

Reply:
> Directory.Exists is already this, is it not? I'm not clear on what this method is for.

The designer needs to handle a couple of corner cases.  For example:

- I want my tree to show X:\ Y:\ and Z:\ directories even though those don't exist on my dev machine
- I want my tree to show the root `/` because I'm dev on Windows but will run my app in a linux docker container

And other use cases.  But you are right, this method is not the way to go.

I think that a better approach than this method is just to:

- Ask the user when adding if they want to choose a file or dir (i.e. before even showing FileDialog).

But secondly when I test with a non existant dir (e.g. J:\) I realize that I need to add a try/catch to the child getter so that it swallows exceptions and just serves Empty when user tries to expand those.

Reply:
> I think that a better approach than this method is just to:
> 
> * Ask the user when adding if they want to choose a file or dir (i.e. before even showing FileDialog).

Yeah, I think the way that most UIs do it is probably a good thing to emulate here, too.\
For example, Windows Explorer, Visual Studio, and Dolphin (Linux) use separate context menu entries for adding different types of items. One they all have in common is that folders are their own entry.



> Thinking more about this I think also I should make it possible to turn off codegen for the `TreeBuilder` property if you want to write your own filesysteminfo child builder e.g. that has some custom logging or exception handling.

I did a bunch of work building trees using V1 of TG for my SnapsInAZfs project, because it needs to display the ZFS hierarchy as a tree, for the most logical layout for users. It also uses that tree to enable cascading event dispatch for when a property is changed at a higher level that is to be inherited (or not) to its descendants.

In essence, the important parts are the nodes contain ordered dictionaries of children, sooooorta-kinda-not-really like a B+ tree. When fed to Terminal.Gui, they are turned into appropriate Lists, and then their Data properties are used to hold references back to the actual more complex nodes.

Ha, come to think of it, there are actually several overlapping concepts in use between this and that application, at least for this area of tree work. Another is the way that `Properties` are handled - in TGD, it's to support properties of an object. In SIAZ, it's to support ZFS Properties (name/value/source tuples, for the most part). Those properties have to follow standard inheritance rules. I made them with a fully generic implementation, as well as an interface. Might be some things we could pilfer from it for use here. 😅 

But back to the paths stuff - yeah, I say make it convenient to a point, but we are dealing with developers, so (at least I hope) we can probably assume at least a basic familiarity with files and folders. 😆